src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataReaderExtensions.cs

```diff
@@ -85,30 +85,31 @@ private static void ReadCollectionItem(ODataReader reader, Stack<ODataItemBase>
                     }
                     else
                     {
-                        ODataItemBase parentItem = itemsStack.Peek();
-                        ODataResourceSetWrapper parentResourceSet = parentItem as ODataResourceSetWrapper;
-                        if (parentResourceSet != null)
-                        {
-                            parentResourceSet.Resources.Add(resourceWrapper);
-                        }
-                        else
-                        {
-                            ODataNestedResourceInfoWrapper parentNestedResource = (ODataNestedResourceInfoWrapper)parentItem;
-                            Contract.Assert(parentNestedResource.NestedResourceInfo.IsCollection == false, "Only singleton nested properties can contain resource as their child.");
-                            Contract.Assert(parentNestedResource.NestedItems.Count == 0, "Each nested property can contain only one resource as its direct child.");
-                            parentNestedResource.NestedItems.Add(resourceWrapper);
-                        }
+                        AddResourceToParent(itemsStack, resourceWrapper);
                     }
 
                     itemsStack.Push(resourceWrapper);
                     break;
 
                 case ODataReaderState.ResourceEnd:
+                case ODataReaderState.DeletedResourceEnd:
                     Contract.Assert(
                         itemsStack.Count > 0 && (reader.Item == null || itemsStack.Peek().Item == reader.Item),
```

Comment:
>reader.Item == null [](start = 49, length = 19)

reader.Item should never be null, right? We should assert that separately, or remove the check from this assert, but it doesn't make sense to have it as part of thi assert.